### PR TITLE
Testing for DatastoreManager

### PR DESCRIPTION
**Pull request by ade-ademiluyi**

### CODE REVIEW

File: src/test/java/com/google/sps/managers/DatastoreManagerTest.java [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+    User.Builder user = new User.Builder("0001").setUsername("user1").setFirstName("Ade").setLastName("Ademiluyi");
+    user.setProfilePictureUrl("images/defaultProfilePicture.jpg").setGamesCreated(gamesCreated);
+    user.setGamesCompletedWithTime(new ArrayList<Pair<String, Long>>());
+    expected = user.build();
+
+    datastoreManager.createOrReplaceUser(expected);
+
+    actual = datastoreManager.retrieveUser("0001");
+
+    Assert.assertEquals(expected.getUserID(), actual.getUserID());
```
Comment (ldcWV) [inline src/test/java/com/google/sps/managers/DatastoreManagerTest.java:88]: Probably also check that the other fields are also preserved, because you already went through the effort of setting all of them.

File: src/test/java/com/google/sps/managers/DatastoreManagerTest.java [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+    Game.Builder game = new Game.Builder("0001", "game1").setGameCreator("Adetowo").setStages(stages);
+    game.setGameDescription("For Testing").setNumTimesPlayed(14).setNumTimesFinished(12);
+    game.setNumStarVotes(6).setTotalStars(17).setNumDifficultyVotes(6).setTotalDifficulty(6);       
+    expected = game.build();
+
+    datastoreManager.createOrReplaceGame(expected);
+
+    actual = datastoreManager.retrieveGame("0001");
+
+    Assert.assertEquals(expected.getGameID(), actual.getGameID());
```
Comment (ldcWV) [inline src/test/java/com/google/sps/managers/DatastoreManagerTest.java:194]: Also check that the other fields are equal here.

File: src/test/java/com/google/sps/managers/DatastoreManagerTest.java [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+    //Makes use of the SinglePlayerProgress class defined previously to build a single SinglePlayerProgress for testing purposes
+    SinglePlayerProgress.Builder progress = new SinglePlayerProgress.Builder("0000", "0001");
+    progress.setLocation(new Coordinates()).setHintsFound(new ArrayList<Integer>()).setStageID("0002");
+    expected = progress.build();
+
+    datastoreManager.createOrReplaceSinglePlayerProgress(expected);
+
+    actual = datastoreManager.retrieveSinglePlayerProgress("0000", "0001");
+
+    Assert.assertEquals(expected.getUserID(), actual.getUserID());
```
Comment (ldcWV) [inline src/test/java/com/google/sps/managers/DatastoreManagerTest.java:316]: Here too.

File: src/test/java/com/google/sps/managers/DatastoreManagerTest.java [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+    //Makes use of the Hint class defined previously to build a single Hint for testing purposes
+    Hint.Builder hint = new Hint.Builder("0001", 3);
+    hint.setLocation(new Coordinates()).setText("For testing");
+    expected = hint.build();
+
+    datastoreManager.createOrReplaceHint(expected);
+
+    actual = datastoreManager.retrieveHint("0001");
+
+    Assert.assertEquals(expected.getHintID(), actual.getHintID());
```
Comment (ldcWV) [inline src/test/java/com/google/sps/managers/DatastoreManagerTest.java:352]: Here too.

File: src/test/java/com/google/sps/managers/DatastoreManagerTest.java [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+    ArrayList<Game> test = new ArrayList<>();
+  
+    for(int i = 0; i < 10; i++) test.add(getRandomGame());
+    actual =  test.size();
+
+    Assert.assertEquals(expected, actual);
+  }
+  
+  @Test
+  public void doesUsernameExistTest() {
```
Comment (ldcWV) [inline src/test/java/com/google/sps/managers/DatastoreManagerTest.java]: Also add a test for when the username doesn't exist.

File: src/test/java/com/google/sps/managers/DatastoreManagerTest.java [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+
+    datastoreManager.createOrReplaceUser(test);
+
+    actual = datastoreManager.doesUsernameExist("user1");
+
+    Assert.assertEquals(expected, actual);
+  }
+
+  @Test
+  public void deleteSinglePlayerProgressTest() {
```
Comment (ldcWV) [inline src/test/java/com/google/sps/managers/DatastoreManagerTest.java:430]: Add a case where you are trying to delete a SinglePlayerProgress that doesn't exist. In this case you should probably throw an error.

File: src/test/java/com/google/sps/managers/DatastoreManagerTest.java [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+
+    if (test2 == null) {
+      actual = false;
+    }
+
+    Assert.assertEquals(expected, actual);
+  }
+  
+  @Test
+  public void retrieveUserByUsername() {
```
Comment (ldcWV) [inline src/test/java/com/google/sps/managers/DatastoreManagerTest.java]: Add a case where there is no user with that username. Then the retrieveUser function should return null.

File: src/test/java/com/google/sps/managers/DatastoreManagerTest.java (hunk 5 above)
Comment (ade-ademiluyi) [reply src/test/java/com/google/sps/managers/DatastoreManagerTest.java]: sure. already wroking on testing edge cases

File: src/test/java/com/google/sps/managers/DatastoreManagerTest.java (hunk 6 above)
Comment (ade-ademiluyi) [reply src/test/java/com/google/sps/managers/DatastoreManagerTest.java:430]: sure. already wroking on testing edge cases

File: src/test/java/com/google/sps/managers/DatastoreManagerTest.java (hunk 7 above)
Comment (ade-ademiluyi) [reply src/test/java/com/google/sps/managers/DatastoreManagerTest.java]: sure. already wroking on testing edge cases

File: src/test/java/com/google/sps/managers/DatastoreManagerTest.java (hunk 1 above)
Comment (ade-ademiluyi) [reply src/test/java/com/google/sps/managers/DatastoreManagerTest.java:88]: sure

**Review by saramissak — APPROVED** [review]

File: src/test/java/com/google/sps/managers/DatastoreManagerTest.java [hunk 8]
```diff
@@ -210,11 +410,21 @@ public void doesUsernameExistTest() {
...
 
     datastoreManager.createOrReplaceUser(test);
+
     actual = datastoreManager.doesUsernameExist("user1");
 
     Assert.assertEquals(expected, actual);
   }
 
+  @Test
+  public void doesUsernameExistTestFalse() {
```
Comment (saramissak) [inline src/test/java/com/google/sps/managers/DatastoreManagerTest.java:421]: Shouldn't this throw an error now or no?

File: src/main/java/com/google/sps/managers/DatastoreManager.java [hunk 9]
```diff
@@ -385,25 +386,39 @@ public User retrieveUserByUsername(String username) {
...
+  public Map<String, String> retrieveKeys()
+  {
+    Map<String, String> apiKeys = new HashMap<>();
+    
+    Query query = new Query("apiKeys");
+    PreparedQuery pq = datastore.prepare(query);
+    Entity userEntity = pq.asSingleEntity();
+
+    String clientId = (String) userEntity.getProperty("CLIENT_ID");
+    String mapApi = (String) userEntity.getProperty("MAP_key");
```
Comment (ldcWV) [inline src/main/java/com/google/sps/managers/DatastoreManager.java]: Should probably capitalize 'key' here to keep things consistent.

**Review by ldcWV — APPROVED** [review]

File: src/main/java/com/google/sps/managers/DatastoreManager.java (hunk 9 above)
Comment (ade-ademiluyi) [reply src/main/java/com/google/sps/managers/DatastoreManager.java]: done

Comment (ade-ademiluyi) [reply src/main/java/com/google/sps/managers/DatastoreManager.java]: done
